### ISSUE TITLE
Training script for LAVA

### ISSUE THREAD

**Issue by MaximilianToelle** [google-research/language-table#5]
Hi guys!

I am a big fan of your work and would like to ask if you have considered also providing the training script for LAVA. Especially the part about action and rgb normalization is currently not clear to me.

Thanks in advance :)

**Comment by jacob-zietek** [comment]
Bumping this

**Comment by ayzaan** [comment]
Hi @MaximilianToelle, I've just pushed a portion of the code for normalization which you can find in this commit: https://github.com/google-research/language-table/commit/2cabececffa3c1420345fd7df522a7664532b2c0
This contains the normalization code which you can take a look at, and I also added the jax BC implementation.

For the full train script and data code, that is a bit delayed since we are waiting on a new version release for one of the dependencies in our data pipeline. Apologies for the delay on this, but hopefully the above code is useful for now. Let me know if you have any questions.

**Comment by berniewang8177** [comment]
Hi,

Thank you for the fantastic work! I really enjoy reading the paper and want to build something upon LAVA !
Could you anticipate a time that the full training / evaluation scripts will released?

Again, great work!

**Comment by MaximilianToelle** [comment]
Thanks @ayzaan for parts of the training code! While looking through the code, some questions came to my mind. It would be awesome if you could provide some more insights :) 

- I don't quite understand the loss calculation. The [per_example_loss](https://github.com/google-research/language-table/blob/2cabececffa3c1420345fd7df522a7664532b2c0/language_table/train/bc.py#L233) seems to be a single value as jax.numpy.mean takes the mean over all elements. This in turn is used to form a mean value to generate the final [loss](https://github.com/google-research/language-table/blob/2cabececffa3c1420345fd7df522a7664532b2c0/language_table/train/bc.py#L182). Shouldn't it be something like `loss = jnp.mean(jnp.sum(jnp.square(predicted_actions - action), axis=1))`
- I've recognized that the MultiScaleResNet Encoder is set to [train=False](https://github.com/google-research/language-table/blob/2cabececffa3c1420345fd7df522a7664532b2c0/language_table/train/networks/lava.py#L60). Why is it nevertheless necessary to include batch_stats as a mutable during training? 
- Do you guys use the downloaded clip checkpoint to encode the instructions within the data pipeline and then use the ["clip_in_obs"](https://github.com/google-research/language-table/blob/2cabececffa3c1420345fd7df522a7664532b2c0/language_table/train/networks/lava.py#L420) option during training?
- Using your evaluation script, I have seen that the instructions are also provided as sequences. The embeddings within a sequence seem to always be the same as the instruction does not change within a single episode. Do you also filter out sequences of changing instructions in your data pipeline? Does training suffer if you would only provide a single instruction embedding instead of a sequence of identical ones?

**Comment by ayzaan** [comment]
The training code is available now!

The python dependency chain was a bit tricky, mainly because of a TF dependency on specific protouf versions. I've added a `requirements_static.txt` with pinned dependency versions that works on machines I've tested on.

```
python3 -m venv ./ltvenvtrain
source ./ltvenvtrain/bin/activate
pip install --no-deps -r ./requirements_static.txt
export PYTHONPATH=${PWD}:$PYTHONPATH
```

Also need to download the CLIP weights:
```
python language_table/train/download_clip_flax_ckpt.py --checkpoint_directory=/tmp/scenic_clip_ckpt
```

Running the training:
```
mkdir -p /tmp/language_table_train/
PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python  python language_table/train/main.py --config=./language_table/train/configs/language_table_sim_local.py --workdir=/tmp/language_table_train/
```

The `language_table_sim_local.py` file is a config that should be good for testing locally. It has a smaller batch size and replay_capacity. Let me know if you have any questions!

**Comment by ayzaan** [comment]
@MaximilianToelle, to answer your questions:

* Ah yes this is wrong and/or bad variable naming, since we're taking the mean it's not a per example loss. I'll fix this, or feel free to submit a PR yourself.
* It's not necessary in this case, since we're freezing the ResNet here. For flexibility, we left it in since in other cases we may want to unfreeze the ResNet for finetuning, other implementations of networks may use batchnorm, etc.
* In this training script we're tokenizing the CLIP instruction in the data pipeline and adding the frozen CLIP text encoder to the network. We've previously done an optimization where we run the CLIP text encoder offline to produce embeddings and bake it into the dataset to avoid the forward pass during training. However we released the public dataset with just raw text instructions rather than including the CLIP embeddings in, so we're not doing the optimization here.
* Apologies, we do use the ViT-B/32 model during the CLIP finetuning procedure. Note that in this BC training setup we're only using the text encoder, so it doesn't really matter which we download. But for consistency with the paper I'll change it, thanks for pointing it out.
* Yes the instruction doesn't change across the episode so passing the sequence of instructions is not necessary. We just arbitrarily take the first item in the sequence in the network: `tokens = x["instruction_tokenized_clip"].astype(jnp.int32)[:, 0]`. It shouldn't affect training at all. In the dataset, each episode will only have a single instruction, so we don't really have the case where there are sequences of changing instructions.

**Comment by MaximilianToelle** [comment]
Thanks a lot for the full training code and your detailed answers @ayzaan, it really helped me to get forward! Some last questions came up and it would be awesome if you can once again answer them :) 

- In the paper you mention the use of "two Imagenet-pretrained ResNet layers". Though, I've recognized that the ResNet layers are being trained and not downloaded as with clip. What would be the correct way for training? 
- In the evaluation script, I've seen the comment "For PUSH we need to normalize RGB". In the config file, I've seen config.data_skip_normalize_keys=["rgb", "instruction"]. At the end, did you normalize the rgbs or not?

**Comment by ayzaan** [comment]
Yes the model we used in real used `lava_image_encoder="resnet"` with this config to freeze parameters:

```
config.agent.pretrained_checkpoints = [
      (
        "<clip checkpoint>",
        [("params/text", "params/encoder/TextEncoder_0")],
      ),
      ("<resnet imagenet checkpoint>",
       [("params", "params/encoder/ResNetVisualEncoder_0/MultiscaleResNet_0"),
        ("batch_stats",
         "batch_stats/encoder/ResNetVisualEncoder_0/MultiscaleResNet_0")])
  ]
  config.agent.freeze_keys = ["TextEncoder_0", "MultiscaleResNet"]
```

This requires a pretrained resnet checkpoint (we pretrained on imagenet classification). I'll see if I can provide this while addressing #9. Overall, the `conv_maxpool` image encoder has very similar performance in sim.

Re: normalization - Those comments were outdated from a non-related task, I removed them in 039dfa4b6b610b431e33ab1db44afa1a18643b54. In the data pipeline we scale to [0.0, 1.0] with `image = tf.cast(image, tf.float32) / 255.0`. For conv_maxpool no extra handling is done. For resnet, we want to normalize the image in the same way as was done during the pretraining, which is "standard" resnet normalization (https://github.com/google-research/language-table/blob/c617df4a4fa88e07b6c83cdac89a4d7eee25c2b8/language_table/train/networks/lava.py#L90). Since this normalization is done in the network, we don't need any special handling in policy.py when doing inference.
